### DISCUSSION TITLE
T-PicoC3 trying communication between RP2040 and ESP32

### DISCUSSION THREAD

**Discussion by lesept777** [micropython/micropython#9212] [Hardware & Peripherals] [answered]
Hi, I just received my new[ T-PicoC3](https://github.com/Xinyuan-LilyGO/T-PicoC3) board, and I'm discovering Micropython as well.
I want to test the communication between the processors.

Basically, the idea is to use a UART communication. But, I'm not sure how to code them on both processors.

1) On the ESP32 side: I write the following code, saved as com_RP.py
```
from machine import UART, Pin
from time import sleep

# RP2040 = UART(1, baudrate=115200, tx=Pin(9), rx=Pin(8), cts=Pin(10), rts=Pin(11), txbuf=1024, rxbuf=1024)
RP2040 = UART(1, baudrate=9600, tx=9, rx=8, cts=10, rts=11, txbuf=1024, rxbuf=1024)
RP2040.init()
print('ok')
sleep(5)
for i in range (15):
    buf = str(i)+'\n'
    RP2040.write(buf)
    print(i)
    sleep(1)
```

It seems that rx, tx, rts, cts arguments must be numbers and not Pin. When I run this code, it takes around 15s before I get the 'ok' message on Thonny's console. It's weird, because if I type them in the console, I get the message almost immediately... 

2) On the RP2040 side; I write the second code, saved as com_ESP32.py
```
from machine import UART, Pin
from time import sleep
import st7789
import tft_config
import NotoSans_32 as font

led = Pin(25, Pin.OUT)
for i in range(10):
    led.value(not led.value())
    sleep(0.05)
led.off()
tft = tft_config.config(1)
tft.init()
bg = st7789.BLUE
fg = st7789.WHITE
tft.fill(bg)
tft.write(font, "Waiting...", 0, 20, fg, bg)
print ("waiting...")
sleep(1)
ESP32 = UART(1, baudrate=9600, tx=Pin(8), rx=Pin(9), cts=Pin(10), rts=Pin(11), txbuf=1024, rxbuf=1024)
ESP32.init()

while True:
    if ESP32.any() > 0:
        line = ESP32.readline()
        print("*"+str(line)+"*")
        if bg == st7789.BLUE:
            bg = st7789.GREEN
        else:
            bg = st7789.BLUE
        if line != '':
            led.value(not led.value())
            tft.fill(bg)
            tft.write(font, line, 0, 20, fg, bg)
        sleep(0.5)
```

It's a little longer, because this processor can address the LED and the display (the ESP32 cannot as far as I know).
In the UART definition, I switched rx and tx, but not rts and cts. Is this ok?
I run both codes at boot by putting their names in a boot.py file, such as, for the ESP32:
`import com_RP`

When the board boots, I see the LED blink, then the display becomes blus and the "Waiting..." message appears on it. Then the screen goes green, the LED switches on meaning that something arrives in the UART, but nothing happens after that.
Can anyone help? Thanks a lot.

**Comment by lesept777** [comment, marked as answer]
From [this discussion](https://github.com/orgs/micropython/discussions/9161), it seems that I have to use different pin numbers for UART1:

> tx must be 0, 4, 8, 12...etc for uart 0, 1, 2, 3...etc
> rx must be 1, 5, 9, 13...etc for uart 0, 1, 2, 3...etc

So I changed in both files. On the ESP32's side:
`RP2040 = UART(1, baudrate=9600, tx=5, rx=4, cts=7, rts=6, txbuf=1024, rxbuf=1024)`
and on the RP2040's side:
`ESP32 = UART(1, baudrate=9600, tx=Pin(4), rx=Pin(5), cts=Pin(6), rts=Pin(7), txbuf=1024, rxbuf=1024)`
(crossing rx/tx and cts/rts) but it's still not working... The display remains blue, with the "Waiting..." message.

**Reply by robert-hh** [reply]
Did you consider, that at RP2040 the numbers for rx, tx, .. are GPIO numbers, not board numbers. e.g. GPIO 4 is at board pin # 6.

**Reply by robert-hh** [reply]
You may also first start without using rts and cts. Bot board have buffered IO. So as long as your messages as shorter than the buffer flow control is not needed.

**Reply by lesept777** [reply]
> Did you consider, that at RP2040 the numbers for rx, tx, .. are GPIO numbers, not board numbers. e.g. GPIO 4 is at board pin # 6.

On this board, it seems that GPIO and pin numbers are identical.
![image](https://user-images.githubusercontent.com/30208898/188442035-e41de822-95af-4ac2-9b3c-8b6eb7349327.png)

> You may also first start without using rts and cts.

I have tried with no success. I think, seeing the schematics above, that there is no internal connection between both processors, and that I need to wire the pins I want to use. For example, if I define on the ESP32 side:
`RP2040 = UART(1, baudrate=9600, tx=9, rx=8)`
and on the RP2040 side:
`ESP32 = UART(1, baudrate=9600, tx=Pin(28), rx=Pin(29))`
then I should add a wire between GPIO8 and GPIO 28, and another between GPIO9 and GPIO29.
How does that sound to you?

**Reply by robert-hh** [reply]
That sounds OK. Looking at the schematics, there are internal connections betweem R>P2040 andd ESP32. GPIO 8-11 are connected to the ESP32, with GPIO
- 8 labeled as TX to ESP32 MTCK (GPIO6)
- 9 as RX to ESP32 MTDO (GPIO7),
- 10 as CTS to ESP32 MTMS (GPIO4) and
- 11 as RTS to ESP32 MTDI (GPIO5)

**Reply by lesept777** [reply]
Thanks a lot. So I should do: 
RP2040 side
`ESP32 = UART(1, baudrate=9600, tx=Pin(8), rx=Pin(9), cts=Pin(10), rts=Pin(11))`
ESP32 side
`RP2040 = UART(1, baudrate=9600, tx=6, rx=7, cts=4, rts=5)`
Is it correct?

**Reply by lesept777** [reply]
> Thanks a lot. So I should do: RP2040 side `ESP32 = UART(1, baudrate=9600, tx=Pin(8), rx=Pin(9), cts=Pin(10), rts=Pin(11))` ESP32 side `RP2040 = UART(1, baudrate=9600, tx=6, rx=7, cts=4, rts=5)` Is it correct?

EDIT: it's not working... :(

**Reply by robert-hh** [reply]
You have to connect tx to rx and rx to tx, so the ESP32 side should be:
`RP2040 = UART(1, baudrate=9600, tx=7, rx=6, cts=5, rts=4)`

**Reply by lesept777** [reply]
Thanks. I tried this but no sign of the expected communication. Maybe the problem is somewhere else.
I thought that the codes are run at boot, thanks to the files boot.py on each side, but maybe it's not like this for the ESP32?
Or my processing of incoming messages in the RP2040 is not correct?
Or maybe I shouldn't use UART number 1? I'm not really familiar with this in Micropython (I usually use C).

**Reply by robert-hh** [reply]
The UART numbers must be 1 for RP2040 and can be 1 for ESP32. 
boot.py is run by both ports, but code like that should better be placed in main.py, which is called after boot.py.
In your ESP32 code, do not call `rp2040.init()`. It is not required and may be bad, because it will set the device back to default settings. The same applies to `ESP32.init()` in the RP2040 code.

**Reply by lesept777** [reply]
Great! Thanks a lot for your help, it works now. Not exactly as I expected, but I can see that there is indeed communication.
I'll work on it to improve it.

**Comment by lesept777** [comment]
For those who would like to use the UART communication between the two processors, here is what I did:

ESP32 side:
```
from machine import UART
RP2040 = UART(1, baudrate=38400, tx=7, rx=6, cts=5, rts=4)
```

RP2040 side:
```
from machine import UART
ESP32 = UART(1, baudrate=38400, tx=Pin(8), rx=Pin(9), cts=Pin(10), rts=Pin(11))
```

To read data from the UART:
```
 if ESP32.any() > 0:
    buf = ESP32.readline()
    line += buf.decode("utf-8")
```

It seems that higher baudrates do not work.
Hope this will help anyone.
